hdate/date.py

```diff
@@ -237,11 +251,60 @@ def upcoming_shabbat(self):
 
         If it is currently Shabbat, returns the HDate of the Saturday.
         """
+        if self.is_shabbat:
+            return self
         # If it's Sunday, fast forward to the next Shabbat.
         saturday = self.gdate + datetime.timedelta(
             (12 - self.gdate.weekday()) % 7)
         return HDate(saturday, diaspora=self.diaspora, hebrew=self.hebrew)
 
+    @property
+    def upcoming_shabbat_or_yom_tov(self):
+        """Return the HDate for the upcoming or current Shabbat or Yom Tov.
+
+        If it is currently Shabbat, returns the HDate of the Saturday.
+        If it is currently Yom Tov, returns the HDate of the first day
+        (rather than "leil" Yom Tov). To access Leil Yom Tov, use
+        upcoming_shabbat_or_yom_tov.previous_day.
+        """
+        if self.is_shabbat or self.is_yom_tov:
+            return self
+
+        if self.upcoming_yom_tov < self.upcoming_shabbat:
+            return self.upcoming_yom_tov
+        return self.upcoming_shabbat
+
+    @property
+    def first_day(self):
+        """Return the first day of Yom Tov or Shabbat.
+
+        This is useful for three-day holidays, for example: it will return the
+        first in a string of Yom Tov + Shabbat.
+        If this HDate is Shabbat followed by no Yom Tov, returns the Saturday.
+        If this HDate is neither Yom Tov, nor Shabbat, this just returns
+        itself.
+        """
+        day_iter = self
+        while day_iter.previous_day.holiday_type == HolidayTypes.YOM_TOV:
```

Comment:
Didn't think this completely through, but shouldn't it be:

`while day_iter.previous_day.holiday_type == HolidayTypes.YOM_TOV or day_iter.previous_day.is_shabbat`?

Reply:
Yes, you're right, fixed.